### DISCUSSION TITLE
Kickoff of cloud native discussion

### DISCUSSION THREAD

**Discussion by planetf1** [odpi/egeria#7323] [Ideas]
At our recent Egeria team face to face, we agreed to start up a work-group on ‘cloud native’.

Egeria’s current service model is based on our [‘server chassis’](https://egeria-project.org/concepts/omag-server-platform/) or platform, within which we host ’[servers’](https://egeria-project.org/concepts/omag-server/?h=omag+server). This is a java process
These servers can be of varying types including [metadata repository](https://egeria-project.org/services/omrs), [integration daemon](https://egeria-project.org/services/omis) servers, [metadata access point](https://egeria-project.org/services/omas), [view services](https://egeria-project.org/services/omvs) (UI support), and others.

These servers then host 1 or more services – be it OMRS (for a repository), an [OMAS](https://egeria-project.org/services/omas/) (ie [Asset Consumer](https://egeria-project.org/services/omas/asset-consumer/overview/#overview) access service), an [OMIS](https://egeria-project.org/services/omis/) (integration service)

The configuration specification is geared around JSON configuration documents. These documents can be stored/provided, or built incrementally by configuration APIs. Additionally management of the services they define I through APIs (ie [Admin Services](https://egeria-project.org/services/admin-services/overview/?h=admin+services) & [Platform services](https://egeria-project.org/services/platform-services/overview/?h=platform+services)         )

Our [current container support](https://egeria-project.org/guides/operations/kubernetes/) which I’ve developed over the last few years (starting from the perils of running a lab at IBM Think!) is mostly oriented around a single container image for the core java process ie platform. We have helm charts that setup 1 or more platforms plus ancillary components – UIs, Kafka etc. These have been designed to support education, demo, tutorials, getting-started, but not production. They have, however, been used by many as a base, including  for production with more work around ci/cd.

We also have a [Kubernetes Operator](https://egeria-project.org/guides/operations/kubernetes/operator/#further-information) with a single CRD that will setup/manage platforms including scaling, but make use of existing configuration documents. This provides active control, but only to the platform level. This hasn’t been released. We were about to look at further control within the limits of the existing configuration management support.

Whilst discussed when the operator work started, recently, and especially at the face to face community meetings in Amsterdam, the subject of better ‘cloud native’ support has come up. This is important for for some community members as a way of running in an enterprise.It may be less important for small deployments. The existing operator work has shown that the configuration/operation management is compromised by our current platform architecture – and also, cloud native is a broader discussion than just an operator.

We are forming a workgroup. One of the early tasks will be to define what we mean by ‘cloud native’ and what are the characteristics we need, plus levels of granularity. At this point it likely mean:

- retaining the existing model as one deployment option but *adding* a better k8s/cloud native option
- provision of finer grained services (granularity tbd – Egeria’s sw architecture/packages/maven modules are already fairly fine, so consider this a recomposition)
- removal of admin/platform services – likely replaced by operator/CRDs
- better integration with other k8s services for monitoring, healthcheck etc
- smaller, agile, but more numerous runtimes (eval spring, microprofile, quarkus etc)
- and lots more we need to work through….

I have arranged the first meeting for 1500 UTC on Wed 25 Jan 2023. If you would like an invite please respond here, but we will continue to update here and through issues in github, and feedback on developer, community & TSC calls.

I suggest opening up specific topic threads on areas for discussion. Looking forward to developing new ideas and solutions together!

**Comment by planetf1** [comment]
I've added a mindmap to facilitate the early discussion at https://planetf1.github.io/notes/cloudnative/

**Comment by juergenhemelt** [comment]
Hi Nigel, the SVG link of the mindmap does not seem to be correct. I get a 404.

**Reply by T4ylan** [reply]
Here is an attempt from another tool: 
[Cloud Native Deployment of Egeria.pdf](https://github.com/odpi/egeria/files/10533917/Cloud.Native.Deployment.of.Egeria.pdf)


Edit: Updated with a better quality.

**Reply by planetf1** [reply]
I've updated the link to the svg now.

The mindmap is as it was at the end of the meeting last week.

**Comment by planetf1** [comment]
[call recording](https://ibm.webex.com/ibm/ldr.php?RCID=3fa2b3204dc11766c17b75fc49790e13) - password **kFcyMcW7**

**Comment by planetf1** [comment]
I've added a workgroup page on the wiki - see https://wiki.lfaidata.foundation/display/EG/Cloud+Native . just a quick version so far

**Comment by planetf1** [comment]
Mandy has added a summary of the discussion face to face in the Jan 2023 project newsletter.
See https://egeria-project.org/release-notes/january-2023/#cloud-native-discussion
